spliit: init at 1.19.1

qvalentin opened the pull request:
<!--
^ Please summarise the changes you have done and explain why they are necessary here ^

For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

Add spliit, a free and Open Source Alternative to Splitwise. A webapp to share expenses with your friends and family.
Upstream homepage: https://spliit.app/

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [25.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) (or backporting [24.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) and [25.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

Fixes https://github.com/NixOS/nixpkgs/issues/392139. 

Additional Info:
- I would also like to contribute a NixOS module, for which NixOS tests could be added, that's why I added no package tests for now.
- To run the app some env vars are required and the app, I would handle this in the module
- The app can be started using `npm run start` in the result dir (should a script be added to `bin` to do this?)

Example for the required env vars:

```sh
export POSTGRES_PRISMA_URL=postgresql://postgres:your_password@localhost:5432
export POSTGRES_URL_NON_POOLING=postgresql://postgres:your_password@localhost:5432
export NEXT_PUBLIC_DEFAULT_CURRENCY_SYMBOL=""

export PRISMA_SCHEMA_ENGINE_BINARY=/nix/store/j0xs986nyyg3q8650irb86pvicxmy24i-prisma-engines-5.22.0/bin/schema-engine
export PRISMA_QUERY_ENGINE_BINARY=/nix/store/j0xs986nyyg3q8650irb86pvicxmy24i-prisma-engines-5.22.0/bin/query-engine
export PRISMA_QUERY_ENGINE_LIBRARY=/nix/store/j0xs986nyyg3q8650irb86pvicxmy24i-prisma-engines-5.22.0/lib/libquery_engine.node
```

Let me know if the module should go in the same PR as the package.

qvalentin committed:
spliit: init at 1.15.0

Apply suggestions from code review

Co-authored-by: Yohann Boniface <edhyjox@gmail.com>

qvalentin commented on pkgs/by-name/sp/spliit/package.nix:
```diff
@@ -0,0 +1,58 @@
+    description = "Web UI for sharing expenses with your friends and family, a free and Open Source Alternative to Splitwise.";
+    homepage = "https://spliit.app/";
+    license = lib.licenses.mit;
+    maintainers = with lib.maintainers; [ ];
```
I was unsure whether I was eligible for it as a new contributor. I will add myself.

qvalentin commented on pkgs/by-name/sp/spliit/package.nix:
```diff
@@ -0,0 +1,58 @@
+
+  preBuild = ''
+    prisma generate
+    cp ./scripts/build.env .env
```
Correct, I will add a comment about it.

antonmosich commented:
Your pull request should target the `master` branch, not the `release-24.11` branch.

qvalentin commented:
> Your pull request should target the `master` branch, not the `release-24.11` branch.

https://github.com/qvalentin/nixpkgs/blob/master/CONTRIBUTING.md#how-to-create-pull-requests says to branch of from a commit matching your NixOS version, but if I target master with that the PR contains way to many commits.

I guess I'll move the local base branch to master.

qvalentin committed:
maintainers: init qvalentin

qvalentin commented on pkgs/by-name/sp/spliit/package.nix:
```diff
@@ -0,0 +1,58 @@
+    cp next.config.mjs $out/
+    cp -r prisma $out/prisma
+    cp -r node_modules $out/
+  '';
```
I'm not to sure about the script I just added, but I guess next requires the working directory to contain the `.next` directory.

Sigmanificient commented on pkgs/by-name/sp/spliit/package.nix:
```diff
@@ -0,0 +1,83 @@
+      startScript = writeShellScriptBin "spliit-start-script" ''
+        set -eu pipefail
+
+        export PRISMA_SCHEMA_ENGINE_BINARY="${prisma-engines}/bin/schema-engine";
+        export PRISMA_QUERY_ENGINE_BINARY="${prisma-engines}/bin/query-engine";
+        export PRISMA_QUERY_ENGINE_LIBRARY="${prisma-engines}/lib/libquery_engine.node";
+        export PRISMA_FMT_BINARY="${prisma-engines}/bin/prisma-fmt";
```
Use `lib.getExe`/`lib.getExe'` instead of manually using `/bin/${name}`.

Sigmanificient commented on pkgs/by-name/sp/spliit/package.nix:
```diff
@@ -0,0 +1,83 @@
+        export PRISMA_QUERY_ENGINE_LIBRARY="${prisma-engines}/lib/libquery_engine.node";
+        export PRISMA_FMT_BINARY="${prisma-engines}/bin/prisma-fmt";
+
+        cd "$(dirname "$0")/.."
```
What about using `makeWrapper` with `--set` and `--chdir` (unsure)?

qvalentin replied:
Works quite well if you add the flags in the correct order.

qvalentin committed:
nixos/spliit: init module

qvalentin commented:
The module is already look quite good: https://github.com/qvalentin/nixpkgs/blob/spliit-module/nixos/modules/services/web-apps/spliit.nix

Should I wait till this PR is merged or add it to the same PR?

Sigmanificient commented:
You can add it to the same pr, there is no issue

qvalentin commented on nixos/modules/services/web-apps/spliit.nix:
```diff
@@ -0,0 +1,177 @@
+        }
+      ];
+
+      preStart = lib.mkIf (cfg.database.passwordFile != null) ''
+        DATABASE_PASSWORD="$(cat ${cfg.database.passwordFile})"
+
+        POSTGRES_PRISMA_URL="postgresql://${cfg.database.user}:$DATABASE_PASSWORD@${cfg.database.name}?host=${cfg.database.hostname}:${toString cfg.database.port}";
+
+        echo -n "POSTGRES_PRISMA_URL=\"$POSTGRES_PRISMA_URL\"" > /run/spliit/.env
+        echo -n "POSTGRES_URL_NON_POOLING=\"$POSTGRES_PRISMA_URL\"" > /run/spliit/.env
```
Do you have a idea, how the POSTGRES_PRISMA_URL variables can be given to the service when using cfg.database.passwordFile.

I'm out of ideas.

qvalentin replied:
Seems like this is not supported by prisma (https://github.com/sfackler/rust-postgres/issues/729).

If there is no feasible way to construct the env variables from both the password file and the options, the user would have to enter the whole connection string in the password file and not use the other cfg.database.* parameters.

qvalentin replied:
Thanks, that seems reasonable.

qvalentin commented on nixos/modules/services/web-apps/spliit.nix:
```diff
@@ -0,0 +1,177 @@
+        User = cfg.user;
+        Group = cfg.group;
+
+        # Hardening
```
TODO

qvalentin commented on nixos/modules/services/web-apps/spliit.nix:
```diff
@@ -0,0 +1,177 @@
+  cfg = config.services.spliit;
+in
+{
+  options.services.spliit = with lib; {
```
TODO: Docs

qvalentin replied:
Mostly done

qvalentin committed:
nixos/spliit: init tests

qvalentin commented:
Hi @Sigmanificient could you take another look? There are only some small things missing till we could merge this.

qvalentin committed:
spliit: refactor installPhase

qvalentin committed:
nixos/spliit: user DynamicUser

qvalentin committed:
nixos/spliit: use LoadCredential for dbpasswordfile

ceedubs reviewed (commented):
Nice!

Does this need to be added to /nixos/modules/module-list.nix?

ceedubs commented:
FYI I tried building this against the latest `master` branch I got an error `Type error: Parameter 'activity' implicitly has an 'any' type`. Sorry, I've lost the scrollback with the specifics, but I wanted to give a heads up that there seem to be some recent changes that interfered with this.

qvalentin committed:
nixos/spliit: add to modules list

qvalentin committed:
chore(spliit): do not use build.env

qvalentin committed:
nixos/spliit: remove mdDoc

qvalentin committed:
nixos/spliit: add manual redirects

qvalentin committed:
spliit: 1.15.0 -> 1.16.0

qvalentin commented:
Hi @james-atkins did your suggested change fix the following error for you?

```
   Collecting page data  .   Collecting page data  ..   Collecting page data  ...DOMException [InvalidCharacterError]: Invalid character
    at atob (node:buffer:1302:13)
    at new qr (/build/spliit/node_modules/@prisma/client/runtime/library.js:115:8540)
    at Sl (/build/spliit/node_modules/@prisma/client/runtime/library.js:118:71)
    at new t (/build/spliit/node_modules/@prisma/client/runtime/library.js:131:4651)
    at 13538 (/build/spliit/.next/server/app/groups/[groupId]/expenses/export/json/route.js:1:5572)
    at t (/build/spliit/.next/server/webpack-runtime.js:1:128)
    at 45047 (/build/spliit/.next/server/app/groups/[groupId]/expenses/export/json/route.js:1:671)
    at t (/build/spliit/.next/server/webpack-runtime.js:1:128)
    at t (/build/spliit/.next/server/app/groups/[groupId]/expenses/export/json/route.js:1:5671)
    at /build/spliit/.next/server/app/groups/[groupId]/expenses/export/json/route.js:1:5704 {
  clientVersion: '6.7.0'
}

   Collecting page data  .> Build error occurred
   Collecting page data  .Error: Failed to collect page data for /groups/[groupId]/expenses/export/json
    at /build/spliit/node_modules/next/dist/build/utils.js:1269:15
    at process.processTicksAndRejections (node:internal/process/task_queues:105:5) {
  type: 'Error'
}
   Collecting page data  .
   
```

This seem related to prisma, but sadly even occurs for me when using the Debian binaries.

qvalentin commented:
Seems like bumping prisma in the upstream is easier, maybe this will get merged: https://github.com/spliit-app/spliit/pull/389

qvalentin committed:
spliit: 1.16.0 -> 1.19.0-pre

uses an unreleased version because the prisma update is required for the
nix build to work

qvalentin committed:
spliit: 1.19.0-pre -> 1.19.0

qvalentin commented:
Upstream has released a fixed version so I would say the PR is finally ready for review again.

qvalentin committed:
spliit: 1.19.0 -> 1.19.1

dotlambda commented on nixos/modules/services/web-apps/spliit.nix:
```diff
@@ -0,0 +1,245 @@
+        User = cfg.user;
+        Group = cfg.group;
+        LoadCredential =
+          lib.optional (cfg.secretFile != null) "env-secrets:${cfg.secretFile}"
```
Are you sure this works? I can't find upstream code that uses `env-secrets`. Isn't `environmentFile` enough?

dotlambda commented on pkgs/by-name/sp/spliit/package.nix:
```diff
@@ -0,0 +1,79 @@
+  ];
+
+  env = {
+    NEXT_TELEMETRY_DISABLED = 1;
```
Don't we want this at run time instead?

dotlambda commented on nixos/modules/services/web-apps/spliit.nix:
```diff
@@ -0,0 +1,208 @@
+        echo -n "POSTGRES_URL_NON_POOLING=\"$POSTGRES_PRISMA_URL\"" > /run/spliit/.env
+      '';
+
+      serviceConfig = {
```
This should be patched in our prisma package. See https://github.com/prisma/prisma/issues/17900.

dotlambda commented on nixos/modules/services/web-apps/spliit.nix:
```diff
@@ -0,0 +1,245 @@
+      type = types.port;
+      default = 3000;
+      description = "Port to listen on.";
+    };
```
Use `settings.PORT` instead. You might want to use `freeformType` to specify some options.
```suggestion
```

dotlambda commented on nixos/modules/services/web-apps/spliit.nix:
```diff
@@ -0,0 +1,245 @@
+      type = types.str;
+      default = "spliit";
+      description = "User account under which Spliit runs.";
+    };
```
We use `DynamicUser`.
```suggestion
```

dotlambda replied:
You're the one who set `DynamicUser = true`. But yes, it should work.

dotlambda commented on nixos/modules/services/web-apps/spliit.nix:
```diff
@@ -0,0 +1,245 @@
+      type = types.str;
+      default = "spliit";
+      description = "Group under which Spliit runs.";
+    };
```
```suggestion
```

qvalentin committed:
spliit: add changelog link

Co-authored-by: dotlambda <github@dotlambda.de>

qvalentin commented on nixos/modules/services/web-apps/spliit.nix:
```diff
@@ -0,0 +1,245 @@
+          A file containing the password corresponding to
+          [](#opt-services.spliit.database.user).
+        '';
+      };
```
So those values should then be set with `environmentFile`?

qvalentin commented on nixos/modules/services/web-apps/spliit.nix:
```diff
@@ -0,0 +1,245 @@
+        ]));
+      default = { };
+      description = ''
+        Environment variables settings for spliit.
```
I guess the PORT env variable is just a next feature.

qvalentin committed:
Apply suggestions from code review

nixos/spliit: remove options

Co-authored-by: dotlambda <github@dotlambda.de>

qvalentin committed:
nixos/spliit: rework options (wip)

robert-elles commented:
Trying your pr i get the following when just enabling spliit:

```
error:
       … while calling the 'head' builtin
         at /nix/store/jbddbzsbfrmgbpwzf8gv1jqz5mrfyvyn-nixpkgs-patched/lib/attrsets.nix:1696:13:
         1695|           if length values == 1 || pred here (elemAt values 1) (head values) then
         1696|             head values
             |             ^
         1697|           else

       … while evaluating the attribute 'value'
         at /nix/store/jbddbzsbfrmgbpwzf8gv1jqz5mrfyvyn-nixpkgs-patched/lib/modules.nix:1118:7:
         1117|     // {
         1118|       value = addErrorContext "while evaluating the option `${showOption loc}':" value;
             |       ^
         1119|       inherit (res.defsFinal') highestPrio;

       … while evaluating the option `system.build.toplevel':

       … while evaluating definitions from `/nix/store/jbddbzsbfrmgbpwzf8gv1jqz5mrfyvyn-nixpkgs-patched/nixos/modules/system/activation/top-level.nix':

       (stack trace truncated; use '--show-trace' to show the full, detailed trace)

       error:
       Failed assertions:
       - The options <option>services.spliit.settings.POSTGRES_PRISMA_URL</option> and <option>services.spliit.settings.POSTGRES_URL_NON_POOLING</option>
               should not be set when <option>services.spliit.database.createLocally</option> is set to true

```

Swarsel commented on nixos/modules/services/web-apps/spliit.nix:
```diff
@@ -0,0 +1,223 @@
+              if cfg.database.createLocally then
+                "postgresql://${cfg.database.user}@${cfg.database.name}/${cfg.database.name}?host=/var/run/postgresql/"
+              else
+                "";
```
```suggestion
                null;
```

dotlambda replied:
```suggestion
            default = "postgresql:///spliit?host=/var/run/postgresql/"
              else
                "";
```

Swarsel commented on nixos/modules/services/web-apps/spliit.nix:
```diff
@@ -0,0 +1,223 @@
+              if cfg.database.createLocally then
+                "postgresql://${cfg.database.user}@${cfg.database.name}/${cfg.database.name}?host=/var/run/postgresql/"
+              else
+                "";
```
```suggestion
                null;
```

dotlambda commented on nixos/modules/services/web-apps/spliit.md:
I think it can be removed entirely as long as the module options are documented well.
You can add a changelog entry to inform about the addition of the module.

dotlambda reviewed (commented):
What's the difference between POSTGRES_PRISMA_URL and POSTGRES_URL_NON_POOLING? Does one override the other and should only one be set by default?

dotlambda commented on nixos/modules/services/web-apps/spliit.nix:
```diff
@@ -0,0 +1,223 @@
+        StateDirectory = "spliit";
+        ExecStart = lib.getExe cfg.package;
+        Restart = "always";
+        LoadCredential = lib.optional (cfg.environmentFile != null) "${cfg.environmentFile}";
```
```suggestion
        EnvironmentFile = cfg.environmentFile;
```
